File: DeepLearningDevelopingKit/src/Json/JsonHandler.py
#**************************************************************************************************#
#                                               Deep Learning Developing Kit                                                    #
#								        		 	          Json Handler   	                                                           #
#                                                   www.tianshicangxie.com                                                         #
#                                      Copyright © 2015-2018 Celestial Tech Inc.                                           #
#**************************************************************************************************#

# modules
import os  
import json


# Class : JsonHandler
# Used for handling Json files, loading and writing to Json files.
# Parsing and dumping data from all kinds of dara sources.
class JsonHandler:
    
    json_data = str()

    def __repr__(self):
        return 'JsonHandler()'

    def __str__(self):
        return 'A Json Handler from DeepLearningDevelopingKit.'

    # Open a common file. Such as plaintext.
    @staticmethod
    def open_file(file_path):
        dir_path = os.path.join(os.path.abspath(os.path.join(os.path.dirname(__file__),'..\\..')))
        file_path_abs = dir_path + file_path
        print(file_path_abs)
        if os.path.exists(file_path_abs):
            with open(file_path_abs) as file:
                file_data = file.read()
            file.close()
            return file_data
        else:
            raise Exception('Invalid file path!')

    # Open a Json file and grab all the data in to a python list.
    @staticmethod
    def open_json(file_path):
        dir_path = os.path.join(os.path.abspath(os.path.join(os.path.dirname(__file__),'..\\..')))
        file_path_abs = dir_path + file_path
        print(file_path_abs)
        if os.path.exists(file_path_abs):
            with open(file_path_abs, 'r', encoding='utf-8') as file:
                file_data = json.load(file)
                return file_data
        else:
            raise Exception('Invalid file path!')

    # Dump serialized python object into a Json file.
    @staticmethod
    def save_json(file_path, file_data):
        dir_path = os.path.join(os.path.abspath(os.path.join(os.path.dirname(__file__),'..\\..')))
        file_path_abs = dir_path + file_path
        print(file_path_abs)
        if os.path.exists(file_path_abs):
            with open(file_path_abs, 'w', encoding='utf-8') as file:
                json.dump(file_data, file, indent=4)
        else:
            raise Exception('Invalid file path!')

    # Open a Json file and cache it in json_data.
    @staticmethod
    def open_with_cache(file_path):
        JsonHandler.json_data = JsonHandler.open_json(file_path)

File: DeepLearningDevelopingKit/src/Json/JsonHandler.cpp
/***************************************************************************************************/
/*                                               Deep Learning Developing Kit                                                   */
/*								        		 			  Json Handler 	                                                              */
/*                                                   www.tianshicangxie.com                                                        */
/*                                      Copyright © 2015-2018 Celestial Tech Inc.                                          */
/***************************************************************************************************/

#include "JsonHandler.h"

JsonHandler::JsonHandler()
{

}

MathLib::Vector<double> JsonHandler::ParseVector(const size_t _index) const
{
	if (_index > documentReadBuffer["size"].GetUint())
		exit(0);
	const unsigned dataSize = documentReadBuffer["size"].GetUint();
	const rapidjson::Value& dataBlock = documentReadBuffer["datablock"];

	const rapidjson::Value& vector = dataBlock[_index];

	unsigned size = vector["size"].GetUint();
	MathLib::Vector<double> resultVec(size);

	const rapidjson::Value& data = vector["data"];
	for (rapidjson::SizeType i = 0; i < data.Size(); i++)
		resultVec(i) = data[i].GetDouble();

	return resultVec;
}

std::vector<MathLib::Vector<double>> JsonHandler::ParseAllVector(void) const
{
	std::vector<MathLib::Vector<double>> tempBuffer;
	const unsigned dataSize = documentReadBuffer["size"].GetUint();
	const rapidjson::Value& dataBlock = documentReadBuffer["datablock"];

	for (size_t i = 0; i < dataSize; i++)
	{
		const rapidjson::Value& vector = dataBlock[i];

		unsigned size = vector["size"].GetUint();
		MathLib::Vector<double> tempVec(size);

		const rapidjson::Value& data = vector["data"];
		for (rapidjson::SizeType i = 0; i < data.Size(); i++)
			tempVec(i) = data[i].GetDouble();

		tempBuffer.push_back(tempVec);
	}
	return tempBuffer;
}

void JsonHandler::OpenFile(const string & _filePath)
{
	
}

void JsonHandler::OpenJson(const string & _filePath)
{
	ifstream infile;
	infile.open(_filePath);
	if (!infile.is_open())
		exit(0);
	string dataTemp;
	while (getline(infile, dataTemp))
	{
		jsonReadBuffer += dataTemp;
	}
	infile.close();
	documentReadBuffer.Parse(jsonReadBuffer.c_str());
}

void JsonHandler::SaveJson(const string & _filePath)
{
	ofstream outfile;
	outfile.open(_filePath);
	if(!outfile.is_open())
		exit(0);
	outfile << stringBuffer.GetString();
	
}

void JsonHandler::AppendVectorsToBuffer(const std::vector<MathLib::Vector<double>> & _vec)
{
	documentWriteBuffer.SetObject();
	rapidjson::Document::AllocatorType& allocator = documentWriteBuffer.GetAllocator();

	documentWriteBuffer.AddMember("class", "vector", allocator);
	documentWriteBuffer.AddMember("size", _vec.size(), allocator);

	rapidjson::Value array1(rapidjson::kArrayType);
	for (auto & vec : _vec)
	{
		rapidjson::Value array1obj(rapidjson::kObjectType);
		array1obj.AddMember("class", "vector", allocator);
		array1obj.AddMember("size", vec.Size(), allocator);

		rapidjson::Value array2(rapidjson::kArrayType);
		for (size_t i = 0; i < vec.Size(); i++)
		{
			array2.PushBack(vec(i), allocator);
		}
		array1obj.AddMember("data", array2, allocator);
		array1.PushBack(array1obj, allocator);
	}
	documentWriteBuffer.AddMember("datablock", array1, allocator);

	rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(stringBuffer);
	documentWriteBuffer.Accept(writer);
}
